services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,414 @@
+  mojom::Geoposition response;
+  response.error_code = mojom::Geoposition_ErrorCode::POSITION_UNAVAILABLE;
+  response.error_message = "Unable to create instance of Geolocation API";
+  return response;
+}
+
+// Note: This method blocks because the call to NewSystemProvider is not
+// asynchronous. Fortunately it happens in the background, so this is safe. As
+// it's pretty unlikely that the presence of the GeoClue2 service changes in a
+// browser session, we cache the result here, so we only calculate this once.
+//
+// The easiest way to determine if a DBus service exists is to try and call a
+// method on it, and see if it fails. For this, I chose
+// |GeoClue2.Manager.GetClient| (this is cached on the GeoClue2 side, so it will
+// be the same client we get when we start our service).
+bool GeoClueAvailable() {
+  static bool checked = false;
+  static bool available = false;
+
+  if (checked) {
+    return available;
+  }
+
+  dbus::Bus::Options options;
+  options.bus_type = dbus::Bus::SYSTEM;
+  options.connection_type = dbus::Bus::PRIVATE;
+  auto bus = base::MakeRefCounted<dbus::Bus>(options);
+
+  dbus::ObjectProxy* proxy =
+      bus->GetObjectProxy(kServiceName, dbus::ObjectPath(kManagerObjectPath));
+
+  dbus::MethodCall call(kManagerInterfaceName, "GetClient");
+  auto response =
+      proxy->CallMethodAndBlock(&call, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT);
+
+  // If the response is |nullptr| then the GeoClue2.Manager does not exist.
+  available = response.get();
+  checked = true;
+
+  // Shutdown this bus - we'll use one on the DBus thread for our actual
+  // provider.
+  bus->ShutdownAndBlock();
+
+  return available;
+}
+
+}  // namespace
+
+struct GeoClueProperties : public dbus::PropertySet {
+  dbus::Property<std::string> desktop_id;
+  dbus::Property<uint32_t> requested_accuracy_level;
+
+  explicit GeoClueProperties(dbus::ObjectProxy* proxy)
+      : dbus::PropertySet(proxy, kClientInterfaceName, base::NullCallback()) {
+    RegisterProperty("DesktopId", &desktop_id);
+    RegisterProperty("RequestedAccuracyLevel", &requested_accuracy_level);
+  }
+
+  ~GeoClueProperties() override = default;
+};
+
+struct GeoClueLocationProperties : public dbus::PropertySet {
+  dbus::Property<double> latitude;
+  dbus::Property<double> longitude;
+  dbus::Property<double> accuracy;
+  dbus::Property<double> altitude;
+  dbus::Property<double> speed;
+  dbus::Property<double> heading;
+
+  explicit GeoClueLocationProperties(dbus::ObjectProxy* proxy)
+      : dbus::PropertySet(proxy, kLocationInterfaceName, base::NullCallback()) {
+    RegisterProperty("Latitude", &latitude);
+    RegisterProperty("Longitude", &longitude);
+    RegisterProperty("Accuracy", &accuracy);
+    RegisterProperty("Altitude", &altitude);
+    RegisterProperty("Speed", &speed);
+    RegisterProperty("Heading", &heading);
+  }
+
+  ~GeoClueLocationProperties() override = default;
+
+  void GetAllWithCallback(base::OnceCallback<void()> on_got_all) {
```

Comment:
`WithCallback` is not necessary here. The method signature is sufficient to indicate that there is a callback